pulser/waveforms.py

```diff
@@ -164,7 +170,7 @@ class CompositeWaveform(Waveform):
         waveforms(Waveform): Two or more waveforms to combine.
     """
 
-    def __init__(self, *waveforms):
+    def __init__(self, *waveforms: Waveform):
```

Comment:
Needs to be `Union` with `Parametrized` and `cast` to `Waveform` too. You should put the `cast` above line 179.